experimental/unicodeset_parser/src/parse.rs

```diff
@@ -1098,41 +1071,36 @@ where
         }
     }
 
-    // parses [0-9a-fA-F]{1..6}
-    fn parse_variable_length_hex(&mut self) -> Result<(String, usize)> {
-        let mut result = String::new();
-        let mut end_offset = 0;
-        while let Some(&(offset, c)) = self.iter.peek() {
-            if result.len() >= 6 || !c.is_ascii_hexdigit() {
-                break;
-            }
-            result.push(c);
-            end_offset = offset;
-            self.iter.next();
-        }
-        if result.is_empty() {
-            let (unexpected_offset, unexpected_char) = self.must_peek()?;
-            return Err(PEK::UnexpectedChar(unexpected_char).with_offset(unexpected_offset));
-        }
-        Ok((result, end_offset))
+    fn parse_hex_digits_into_char(&mut self, min: usize, max: usize) -> Result<(usize, char)> {
+        let first_offset = self.must_peek_index()?;
+        let end_offset = self.validate_hex_digits(min, max)?;
+
+        // safety: validate_hex_digits ensures that chars (including the last one) are ascii hex digits,
+        // which are all exactly one UTF-8 byte long, so slicing on these offsets always respects char boundaries
+        #[allow(clippy::indexing_slicing)]
+        let hex_source = &self.source[first_offset..=end_offset];
```

Comment:
thought: if you make `max` a const-generic, you can keep the digits you've seen on the stack and don't need to hold on to `source`. Alternatively, you can merge this with `validate_hex_digit` and calculate the value while iterating with something like `curr = curr * 16 + c.to_digit(16)?`

Reply:
How would the version that keeps digits on the stack work? IIUC, if I want to use a std parsing function I need a UTF-8 slice, but my iterator only gives me `char`s. 

Do you think holding onto `source` is a big downside? If I understand correctly, even in the #3550 - polished version of this API the user will never have to deal with a struct that holds onto `source` (`USParser::parse` takes source and returns the CPILASL)

In any case, if we want to avoid the two-pass approach, I think manually implementing the hex parsing and doing it on the fly as you suggested is ok

Reply:
Yeah the stack based approach is not great, you'd have a `[u8; MAX]` buffer that you'd unsafely interpret as a `str`. Holding on to `source` seems fine